Feature/localization status report

msmcconnell opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description

<!--- Describe your changes in detail -->
This PR adds a LocalizationStatusReport message to describe the status of the localization system for use by the UI and other components. The reported states are based off the state diagram below which is designed to support the following use cases:
* Lidar based localization using NDT matching
* Manual initialization of lidar based localiztion
* Automated initialization of lidar base localization
* Lidar sensor failure transition to GPS based localization
![LocalizationStatusStates](https://user-images.githubusercontent.com/10292079/89468663-f74dae80-d745-11ea-9ce3-2a55c88e51da.png)

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/usdot-fhwa-stol/carma-platform/issues/800
## Motivation and Context
Support UI displays for localization status
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
Builds on CI
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

maefromm commented:
1) Where can I find the fitness score for fault and degraded threshold? What the difference and why do we need to 2 separate threshold for fault and degraded? Maybe this could be explained in the design document, related to node where this may be part of? 

2) For readability, could you put & instead of "and" and maybe have each condition under a new line if it fits? 

2) For INITIALIZING, you have arrow going out saying "fitness score above degraded threshold and auto initialization time exceeded", to be clearer could you say/add "**NDT pose failed to published & **, fitness score above degraded threshold, & auto initialization time exceeded"? 
- Depending on your answer on this and question below, I think your just basically saying "NDT posed failed to publish and auto initialization time exceeded, **regardless** of the fitness score". 

3) For INITIALIZING, what happens  if "**NDT pose failed to publish  &** fitness score **below** degraded threshold, & auto initialization time limit has not been exceeded? 

4) For INITIALIZING, the loop arrow pointing to itself says "fitness score above degraded status", could i suggest to change to "Awaiting for NDT pose to publish while fitness score above threshold"? 

4) What is the initialization time limit, I assume it'll be configuration file that we can change without having to rebuild? 

5) For all the states, you have an arrow saying "initial_pose received" that goes back to INITIALIZING, why is that? 

6) For your NOTE, you said "Smaller values of fitness score are good, while higher values of publication frequency are good". Depending on your answer on #1, for the note, I would like to suggest a legend on the side (doesn't need to be pointing to anything) to explain what below/above or high/low means. Or debating we should not have to say above/below threshold, and just say "fitness score good/bad" to be more straight forward, and not have to explain the threshold? 
- Below Degraded Threshold: Good (and vice versa) 
- Below Fault Threshold: Good (and vice versa) 
- Higher NDT Pose Publication Frequency: Good  (and vice versa) 

7) Between OPERATION -> DEGRADED, the bottom arrow says "NDT pose **not** published at frequency **above** degraded threshold". To be more straight forward, could we refrain using "NOT" word. Maybe say "NDT pose published below the minimum frequency"  

8) Same with "NDT pose published at frequency above degraded threshold and fitness score below degraded threshold". Maybe say "NDT pose published above the minimum frequency & fitness score below degraded threshold" 

9) Under DEGRADED, "NDT pose **not** published at minimum frequency", update to "NDT pose published below minimum frequency" 

10) For AWAIT_MANUAL_INITIALIZATION, I assume it'll stay there until someone takes action or the platform is RESTARTED? Will the user need to restart the platform to re-initialize automatically? Or will this process be triggered again when a new route is selected?

msmcconnell commented:
@maefromm Here is an updated diagram and my responses based on our discussions
![LocalizationStatusStates](https://user-images.githubusercontent.com/10292079/89566158-a5f8fa00-d7ed-11ea-85fd-6bc3b2f3c824.png)

1. As discussed thresholds will be configurable to create the desired ranges. Testing by Shuwei suggested that the fault threshold of Fitness score should be 2.0
2. Changed and to & but it seems I can't put it on a new line with EA
3. As discussed NDT will always publish regardless of the score. So the score and frequency are better indicators.
4. Updated while accounting for 3
5. Yes all thresholds/timeouts will be configurable
6. As discussed the initial pose being published is how the user reinitializes autoware NDT matching. We should allow this to be done from any state to support manual overrides for testing. 
7. Tried to do this with pictures on numberlines. Let me know if that is clearer. 
8/9/10. Removed all negation phrasing
11. Added note in upper right. At the moment I think re-auto initialization will need to be handled by a new yet to be constructed component that can drive this state machine by publishing to initial pose.

msmcconnell commented:
@maefromm Will do
